### PR TITLE
frr: Add new module - frr_bgp

### PR DESCRIPTION
**Pull request by NilashishC**
- Added new module - frr_bgp which provides configuration management of global BGP parameters on devices running Free Range Routing
- Added init files for FRR repo
- Targets - https://github.com/ansible-network/network-proposals/issues/5

### CODE REVIEW

**Comment by NilashishC** [comment]
~Please hold merging this PR, till the directory structure for `module_utils` is updated.~
Holding on to see if there's a way to re-use the docs for top-level parameters that are also sub-options under some other top-level parameter (e.g.: `neighbors` and `networks` which are also a valid sub-option under `address_families`).

**Comment by NilashishC** [comment]
Refactored the existing BGP work to use the CliProvider model.

File: library/frr_bgp.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+      - Replace will remove the existing BGP configuration on the device and re-configure it with the options specified.
+      - Delete will remove the existing BGP configuration from the device.
+    default: merge
+    choices:
+      - merge
+      - replace
+      - delete
+"""
+
+EXAMPLES = """
```
Comment (samccann) [inline library/frr_bgp.py:292]: Same comments here as from the bgp review - IP addresses for documentation should avoid the 1.1.1.1 etc and instead use:
192.0.2.0/24 
198.51.100.0/24
203.0.113.0/24 

And AS numbers for documentation are in the 64496 - 64511 range. 
Should be applied to these examples and in the sample RETURN block

Comment (NilashishC) [reply library/frr_bgp.py:292]: Ack @samccann

**Review by samccann — APPROVED** [review]

File: library/frr_bgp.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+              - multicast
+              - labeled-unicast
+            default: unicast
+          redistribute:
+            description:
+              - Specifies the redistribute information from another routing protocol.
+            suboptions:
+              protocol:
+                description:
+                  - Specifies the protocol for configuring redistribute information.
```
Comment (samccann) [inline library/frr_bgp.py:170]: I noticed in one of the code files here that ospfv3 isn't included in the redistribute list.  It makes me think maybe here and in that other PR (#85) we may need to specify the choices allowed? The other PR definitely had ospfv3. (Unless these protocol options are obvious to the users ...)

Comment (NilashishC) [reply library/frr_bgp.py:170]: Good catch. I'll add that.

**Comment by NilashishC** [comment]
@samccann fixed!

File: library/frr_bgp.py [hunk 3]
```diff
@@ -162,10 +162,24 @@
...
             description:
               - Specifies the redistribute information from another routing protocol.
+              options
             suboptions:
               protocol:
                 description:
                   - Specifies the protocol for configuring redistribute information.
+                choices:
+                  - ospf
+                  - osfp6
```
Comment (samccann) [inline library/frr_bgp.py]: is it really osfp6? or did you mean ospf6??

Comment (NilashishC) [reply library/frr_bgp.py]: Ah. Bad mistake.

File: module_utils/network/frr/providers/cli/config/bgp/__init__.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import re
+
+from ansible.module_utils.six import iteritems
+from ansible.module_utils.network.common.utils import to_list
+from ansible.module_utils.network.frr.providers.providers import register_provider
+from ansible.module_utils.network.frr.providers.providers import CliProvider
+from ansible.module_utils.network.frr.providers.cli.config.bgp.neighbors import Neighbors
+from ansible.module_utils.network.frr.providers.cli.config.bgp.address_family import AddressFamily
+
+REDISTRIBUTE_PROTOCOLS = frozenset(['ospf', 'osfp6', 'eigrp', 'isis', 'table',
```
Comment (samccann) [inline module_utils/network/frr/providers/cli/config/bgp/__init__.py]:
```suggestion
REDISTRIBUTE_PROTOCOLS = frozenset(['ospf', 'ospf6', 'eigrp', 'isis', 'table',
```

**Review by samccann — APPROVED** [review]
docs LGTM

**Comment by NilashishC** [comment]
_Merging these BGP PRs under ansible-network/{{ provider }} are on hold till further discussions_

**Comment by NilashishC** [comment]
Closing this PR since frr_bgp module is already merged in Ansible 2.8.

**Pull request closed without merging**